sidecar/dns/server.go

```diff
@@ -0,0 +1,288 @@
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package dns
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/amalgam8/amalgam8/registry/client"
+	"github.com/miekg/dns"
+)
+
+// Server represent a DNS server. has config field for port,domain,and client discovery, and the DNS server itself
+type Server struct {
+	config    Config
+	dnsServer *dns.Server
+}
+
+// Config represents the DNS server configurations.
+type Config struct {
+	DiscoveryClient client.Discovery
+	Port            uint16
+	Domain          string
+}
+
+// NewServer creates a new instance of a DNS server with the given configurations
+func NewServer(config Config) (*Server, error) {
+	err := validate(&config)
+	if err != nil {
+		return nil, err
+	}
+	s := &Server{
+		config: config,
+	}
+
+	// Setup DNS muxing
+	mux := dns.NewServeMux()
+	mux.HandleFunc(config.Domain, s.handleRequest)
+
+	// Setup a DNS server
+	s.dnsServer = &dns.Server{
+		Addr:    fmt.Sprintf(":%d", config.Port),
+		Net:     "udp",
+		Handler: mux,
+	}
+
+	return s, nil
+}
+
+// ListenAndServe starts the DNS server
+func (s *Server) ListenAndServe() error {
+	logrus.Info("Starting DNS server")
+	err := s.dnsServer.ListenAndServe()
+
+	if err != nil {
+		logrus.WithError(err).Errorf("Error starting DNS server")
+	}
+
+	return nil
+}
+
+// Shutdown stpos the DNS server
+func (s *Server) Shutdown() error {
+	logrus.Info("Shutting down DNS server")
+	err := s.dnsServer.Shutdown()
+
+	if err != nil {
+		logrus.WithError(err).Errorf("Error shutting down DNS server")
+	} else {
+		logrus.Info("DNS server has shutdown")
+	}
+
+	return err
+}
+
+func (s *Server) handleRequest(w dns.ResponseWriter, request *dns.Msg) {
+	response := new(dns.Msg)
+	response.SetReply(request)
+	response.Extra = request.Extra
+	response.Authoritative = true
+	response.RecursionAvailable = false
+
+	for i, question := range request.Question {
+		err := s.handleQuestion(question, request, response)
+		if err != nil {
+			logrus.WithError(err).Errorf("Error handling DNS question %d: %s", i, question.String())
+			// TODO: what should the dns response return ?
+			break
+		}
+	}
+	err := w.WriteMsg(response)
+	if err != nil {
+		logrus.WithError(err).Errorf("Error writing DNS response")
+	}
+}
+
+func (s *Server) handleQuestion(question dns.Question, request, response *dns.Msg) error {
+
+	switch question.Qclass {
+	case dns.ClassINET:
+	default:
+		response.SetRcode(request, dns.RcodeServerFailure)
+		return fmt.Errorf("unsupported DNS question class: %v", dns.Class(question.Qclass).String())
+	}
+
+	switch question.Qtype {
+	case dns.TypeA:
+	case dns.TypeAAAA:
+	case dns.TypeANY:
+	default:
+		response.SetRcode(request, dns.RcodeServerFailure)
```

Comment:
I know that in Golang, you don't need to have expressions below a case statement and that there is no automatic fall through. But that said, for the sake of readability, it would be much nicer if you hoisted the `s.retrieveServices` into the switch case.


Secondly, given that the code below is only supporting TYPEA and TYPEAAA, you should not be having a TYPE_ANY here. ANY is a very bad idea as well. Folks like CloudFlare have deprecated it.

Finally, given that you are retrieving multiple service instances, supporting DNS_SRV seems like a no brainer.